File: tests/homeRefresh.test.ts
import * as assert from 'node:assert/strict';
import { refreshHomeScreenData } from '../legacy/expo-rn/src/utils/homeRefresh';

async function main() {
  const calls: string[] = [];

  let resolveWeather!: () => void;
  let resolveCount!: () => void;

  const runPromise = refreshHomeScreenData(
    {
      fetchWeather: async () => {
        calls.push('weather:start');
        await new Promise<void>((resolve) => {
          resolveWeather = () => {
            calls.push('weather:done');
            resolve();
          };
        });
      },
      fetchTodayStatus: async () => {
        calls.push('status');
      },
      fetchFeedbackCount: async () => {
        calls.push('count:start');
        await new Promise<void>((resolve) => {
          resolveCount = () => {
            calls.push('count:done');
            resolve();
          };
        });
      },
      fetchTodayPrediction: async () => {
        calls.push('prediction');
      },
    },
    { lat: 37.5665, lng: 126.978, force: true }
  );

  await Promise.resolve();
  assert.deepEqual(
    calls,
    ['weather:start', 'status', 'count:start'],
    '예측 계산은 날씨/카운트 준비 전까지 시작되면 안 된다'
  );

  resolveCount();
  await Promise.resolve();
  assert.deepEqual(
    calls,
    ['weather:start', 'status', 'count:start', 'count:done'],
    '카운트만 끝나도 예측은 아직 기다려야 한다'
  );

  resolveWeather();
  await runPromise;
  assert.deepEqual(
    calls,
    ['weather:start', 'status', 'count:start', 'count:done', 'weather:done', 'prediction'],
    '예측 계산은 최신 날씨와 카운트 반영 뒤에 실행되어야 한다'
  );

  console.log('homeRefresh test passed');
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: apps/android-native/app/src/test/java/com/wxxtae/pws/nativepreview/domain/HomeRefreshTest.kt
package com.wxxtae.pws.nativepreview.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class HomeRefreshTest {
    @Test
    fun refreshHomeScreenDataRefreshesPrerequisitesBeforePrediction() {
        val calls = mutableListOf<String>()
        val deps = object : HomeRefreshDeps {
            override fun fetchWeather(lat: Double, lng: Double, force: Boolean) {
                calls += "weather:$lat:$lng:$force"
            }

            override fun fetchTodayStatus() {
                calls += "todayStatus"
            }

            override fun fetchFeedbackCount() {
                calls += "feedbackCount"
            }

            override fun fetchTodayPrediction() {
                calls += "prediction"
            }
        }

        HomeRefreshCoordinator.refreshHomeScreenData(
            deps = deps,
            params = HomeRefreshParams(lat = 37.5665, lng = 126.9780, force = true),
        )

        assertEquals(
            listOf("weather:37.5665:126.978:true", "todayStatus", "feedbackCount", "prediction"),
            calls,
        )
    }
}
